torchtitan/experiments/qwen3/model/state_dict_adapter.py

```diff
@@ -29,58 +31,180 @@ def __init__(self, model_args: Qwen3ModelArgs, hf_assets_path: str | None):
+        2. Split the GroupedExperts' weight into separate expert's wegiht.
+        """
         to_hf_map = {v: k for k, v in self.from_hf_map.items()}
         hf_state_dict = {}
 
         for key, value in state_dict.items():
-            if "layers" in key:
+            if "moe.experts" in key:
                 abstract_key = re.sub(r"(\d+)", "{}", key, count=1)
+                if abstract_key not in to_hf_map:
+                    continue
                 layer_num = re.search(r"\d+", key).group(0)
-                new_key = to_hf_map[abstract_key]
-
-                if new_key is None:
+                new_abstract_key = to_hf_map[abstract_key]
+
+                # Store the GroupedExperts Weight metadata for from_hf()
+                if isinstance(value, DTensor):
+                    self.grouped_expert_weight_placements[abstract_key] = (
+                        value.placements
+                    )
+                    self.grouped_expert_weight_shape[abstract_key] = value.shape
+
+                    # Split GroupedExperts weight to local individual expert weights
+                    local_expert_fqn = self._get_local_experts_weights(
+                        new_abstract_key,
+                        abstract_key,
+                        layer_num,
+                        value,
+                    )
+                    hf_state_dict.update(local_expert_fqn)
+
+                else:
+                    # keep this path for offline conversion
+                    split_values = self._split_experts_weights(
+                        value, self.model_args.moe_args.num_experts
+                    )
+
+                    for expert_num in range(0, self.model_args.moe_args.num_experts):
+                        new_key = new_abstract_key.format(layer_num, expert_num)
+                        hf_state_dict[new_key] = split_values[expert_num].squeeze()
+
+            elif "layers" in key:
+                abstract_key = re.sub(r"(\d+)", "{}", key, count=1)
+                if abstract_key not in to_hf_map:
                     continue
+                layer_num = re.search(r"\d+", key).group(0)
+                new_key = to_hf_map[abstract_key]
                 new_key = new_key.format(layer_num)
+                hf_state_dict[new_key] = value
+
             else:
+                if key not in to_hf_map:
+                    continue
                 new_key = to_hf_map[key]
+                hf_state_dict[new_key] = value
 
-            hf_state_dict[new_key] = value
-
-        return hf_state_dict
+        # Prepare for dequantization
+        hf_state_dict_with_scale_inv = self._add_quantization_scale_inv_tensors(
+            hf_state_dict
+        )
+        return hf_state_dict_with_scale_inv
 
     def from_hf(self, hf_state_dict: dict[str, Any]) -> dict[str, Any]:
+        """
+        1. When loading from HF checkpoint, dequantize the weights from float8 to float32.
```

Comment:
Qwen3 don't need to dequantize, because the weights  for Qwen3 are not quantized